Create Message for Staking Conversion

mattverse opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #XXX

## What is the purpose of the change
Closes:  https://github.com/osmosis-labs/osmosis/issues/5898
 WIP
 
## Testing and Verifying

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

sunnya97 commented on proto/osmosis/superfluid/tx.proto:
```diff
@@ -231,4 +236,35 @@ message MsgAddToConcentratedLiquiditySuperfluidPositionResponse {
+
+  uint64 lock_id = 1 [ (gogoproto.moretags) = "yaml:\"lock_id\"" ];
+  string sender = 2 [ (gogoproto.moretags) = "yaml:\"sender\"" ];
+  string val_addr = 3;
```
We should just use the valset preference module.

mattverse replied:
What if person who's staking does not have val set pref set?

sunnya97 replied:
Then the tx should just fail

sunnya97 replied:
(if they have any stake already, then valset prefs module, fallsback to their current distribution.  so it will only fail if they don't have any staked OSMO already AND they haven't set a preference).

ValarDragon replied:
So is the FE supposed to set valset pref + migrate?

If valset pref is unset, and you've delegated to too many validators, it may also fail

mattverse replied:
proposal: how about we have the val address field as fall back where it would be used as fallback when staking through valset failed?

sunnya97 replied:
mm okay. As long as that's documented.  In the frontend, we will probably not support that tho, to keep it simple

p0mvn replied:
Let's document this decision in this PR please

mattverse replied:
Changed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/0898bdc25c91d561e5f85dac87b4e3dca69431f4

sunnya97 commented on proto/osmosis/superfluid/tx.proto:
```diff
@@ -231,4 +236,35 @@ message MsgAddToConcentratedLiquiditySuperfluidPositionResponse {
+  string sender = 2 [ (gogoproto.moretags) = "yaml:\"sender\"" ];
+  string val_addr = 3;
+  // min_amt_to_stake indicates the minimum amount to stake after conversion
+  string min_amt_to_stake = 4 [
```
Is this acting like a slippage bound basically?

czarcas7ic commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +658,143 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+func (k Keeper) convertGammSharesToOsmoAndStake(
+	ctx sdk.Context,
+	sender sdk.AccAddress, valAddr string,
+	poolIdLeaving uint64, exitCoins sdk.Coins, minAmtToStake sdk.Int,
```
I see no issue (at least for MVP) to use the same pool the lock originated in

czarcas7ic replied:
Would create an issue to eventually do the latter though

AlpinYukseloglu replied:
For routing, one option would be to do a first-order check against the current pool + the linked CL pool if it exists and go with whichever has deeper liquidity. I think adding params that frontends can specify is an appropriate alternative as well (probably in future PR)

mattverse commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+				return
+			}
+			s.Require().NoError(err)
+
+			// Staking & Delegation check
+			// check if old delegation is succesfully deleted
+			_, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, originalValAddr)
+			s.Require().False(found)
+			// check if delegation amount matches
+			delegation, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, valAddr)
+			s.Require().True(found)
+			s.Require().Equal(delegation.Shares, sharesDelegated)
+			s.Require().True(totalAmtConverted.ToDec().Equal(delegation.Shares))
+
+			// Superfluid check
+			// The synthetic lockup should be deleted.
+			_, err = s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, lock.ID, keeper.StakingSyntheticDenom(lock.Coins[0].Denom, valAddr.String()))
+			s.Require().Error(err)
+
+			// Lock check
+			_, err = s.App.LockupKeeper.GetLockByID(s.Ctx, lock.ID)
+			s.Require().Error(err)
+
+			// Bank check
+			balanceAfterConvertLockToStake := s.App.BankKeeper.GetAllBalances(s.Ctx, sender)
+			s.Require().True(balanceBeforeConvertLockToStake.IsEqual(balanceAfterConvertLockToStake))
+		})
+	}
+}
```
I know there's alot of duplication of these validation checks after testing these methods, but this was intentional due to 
- test cases differs a little for each cases
- code duplication  < visibility for test cases 
but open to other opinions

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+		if err != nil {
+			return sdk.ZeroInt(), sdk.ZeroDec(), err
+		}
+		if len(delegationPref.Preferences) != 1 {
```
What do folks think about this? Would this be ok? If val set pref returns multiple validators, made it so that it just errored here cc @sunnya97

ValarDragon replied:
The flow imo should be:
- If valAddr provided, single delegate
- If valAddr not provided, and valset exists, valsetpref.Delegate
- If valAddr not provided, valset does not exist, and superfluid'd, use superfluid validator
- Else: Error

ValarDragon commented:
devbot fix

ValarDragon commented on app/keepers/keepers.go:
```diff
@@ -443,6 +438,11 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
 
 	appKeepers.ValidatorSetPreferenceKeeper = &validatorSetPreferenceKeeper
 
+	appKeepers.SuperfluidKeeper = superfluidkeeper.NewKeeper(
```
Lol oof, were really going to have to come back to superfluid at some point and clean it up. Poor module becoming so omnibus

mattverse replied:
yeah :( poor superfluid keeper :(

ValarDragon commented on proto/osmosis/superfluid/tx.proto:
```diff
@@ -231,4 +236,42 @@ message MsgAddToConcentratedLiquiditySuperfluidPositionResponse {
+  ];
+  // shares_to_convert_and_stake indicates amount of gamm shares to convert and
+  // stake.
+  cosmos.base.v1beta1.Coin shares_to_convert_and_stake = 5 [
+    (gogoproto.moretags) = "yaml:\"shares_to_convert_and_stake\"",
+    (gogoproto.nullable) = false
+  ];
```
Oh wow, were supporting partial's here

czarcas7ic replied:
As talked about on slack, lets go with full conversion only for now

mattverse replied:
yeah, this field was originally added with the intention of 

- being able to add partial logic easier
- supporting partial stake of liquid lp tokens. 

But on the second thought, deleting this field for now

ValarDragon commented on x/superfluid/types/msgs.go:
```diff
@@ -392,3 +393,31 @@ func (msg MsgAddToConcentratedLiquiditySuperfluidPosition) GetSigners() []sdk.Ac
+	}
+
+	return nil
+}
```
We should be doing more formatting validation right?
e.g:
* val_addr either empty or valoper
* slippage bound non-negative

czarcas7ic replied:
This still needs to be addressed, along with respective validate basic test

p0mvn replied:
valoper still needs to be validated right?

mattverse replied:
Right, good catch. Also, I don't think we need a val address validation here since we also do val address validation within the logic (if used)

ValarDragon reviewed (dismissed):
Requested changes for relaxing valset delegation

AlpinYukseloglu commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	var nonOsmoCoins sdk.Coins
+	bondDenom := k.sk.BondDenom(ctx)
+
+	// from the exit coins, separate non-bond denom and bond denom.
```
This is a broader question about the feature: is the intention that we only support OSMO-paired pools (i.e. ones that can be superfluid staked)? It seems as though we never check for one of the denoms to be uosmo or for the pool to be superfluid enabled

I don't see any immediate issues with allowing this feature on all balancer pools, but still asking because it's a deviation from earlier discussions with @czarcas7ic

czarcas7ic replied:
I think it gets indirectly checked when we attempt to swap the denom for osmo. If the pool doesn't contain osmo, this swap will fail.

mattverse replied:
yes, ^ this is the case

czarcas7ic commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+		expectedError bool
+	}
+	testCases := map[string]tc{
+		"happy case": {},
```
can we define what the happy case actually is? I would argue there is no happy case, just multiple cases that test each branch of this method

czarcas7ic commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+	}
+	testCases := map[string]tc{
+		"happy case": {},
+		"use val set preference (single validator)": {
+			useValSetPrefSingleVal: true,
+		},
+		"error: multiple validator returned from valset pref": {
+			useValSetPrefMultipleVal: true,
+			expectedError:            true,
+		},
```
I think we will allow multiple validators after reading the comments, so this will need to get changed as well.

mattverse replied:
Test cases changed in https://github.com/osmosis-labs/osmosis/commit/0898bdc25c91d561e5f85dac87b4e3dca69431f4

czarcas7ic commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
 	}
```
Can we change all these expectedError boolean cases to match errors?

mattverse commented on .golangci.yml:
```diff
@@ -9,7 +9,6 @@ linters:
   enable:
     - asciicheck
     - bidichk
-    - depguard
```
Don't know if it's just for me, but my golang linter started going crazy with dep-guard (import 'github.com/cosmos/cosmos-sdk/store/types' is not allowed from list 'Main' (depguard)go-golangci-lint)

p0mvn replied:
Tbh would prefer if we get to the bottom of the issue / discuss this in a separate PR to avoid blocking this.

Not having any issues myself. Thinking that if we disable a linter every time someone has a local error, the linter will stop being valuable!

mattverse replied:
no worries, if dep guard was my local issue, happy to put it back in

mattverse commented on proto/osmosis/superfluid/tx.proto:
```diff
@@ -231,4 +236,42 @@ message MsgAddToConcentratedLiquiditySuperfluidPositionResponse {
+  ];
+  // shares_to_convert_and_stake indicates amount of gamm shares to convert and
+  // stake.
+  cosmos.base.v1beta1.Coin shares_to_convert_and_stake = 5 [
```
yep, deleting this field for now

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+		}
+	}
+
+	if migrationType == MigrationType(0) || migrationType == MigrationType(1) || migrationType == MigrationType(2) {
```
Good call! Changed

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+		}
+	}
+
+	if migrationType == MigrationType(0) || migrationType == MigrationType(1) || migrationType == MigrationType(2) {
+		totalAmtConverted, totalSharesDelegated, err = k.convertLockToStake(ctx, senderAddr, valAddr, lockID, sharesToConvertAndStake, minAmtToStake)
+		if err != nil {
+			return sdk.ZeroInt(), sdk.ZeroDec(), err
+		}
+	} else if migrationType == MigrationType(3) {
```
Resolved as well!

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	}
+
+	// if superfluid bonded, first change it into superfluid undelegate to burn minted osmo and instantly undelegate.
+	if migrationType == MigrationType(0) {
```
Resolved as well!

mattverse commented on x/superfluid/keeper/migrate.go:
```diff
@@ -332,10 +345,5 @@ func (k Keeper) validateSharesToMigrateUnlockAndExitBalancerPool(ctx sdk.Context
 	}
 
-	// Defense in depth, ensuring we are returning exactly two coins.
-	if len(exitCoins) != 2 {
```
not necessarily! "Force Unlock" handles both superfluid locks and non superfluid locks!

mattverse replied:
But changed to have this check back conditionally regardless

p0mvn replied:
I'm slightly confused why we have this method in the superfluid module then?

mattverse replied:
I think we can make it to be in the lockup keeper, just that more refactoring needs to be done since we also have migration code depending on this method (happy to address this in future code clean up or furture PRs)

czarcas7ic replied:
@ Roman's comment, wouldn't something that modifies both sf and non sf locks belong in sf? SF module has both SF keepers and lockup keepers, while lockup module does not have access to SF keepers.

p0mvn replied:
Makes sense. I'm good with the current approach

mattverse commented on x/superfluid/keeper/migrate.go:
```diff
@@ -97,21 +97,24 @@ func (k Keeper) migrateSuperfluidBondedBalancerToConcentrated(ctx sdk.Context,
 		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, err
 	}
 
+	// Defense in depth, ensuring we are returning exactly two coins.
+	if len(exitCoins) != 2 {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, 0, types.TwoTokenBalancerPoolError{NumberOfTokens: len(exitCoins)}
+	}
```
Good call, changed to be conditional. Also it should work on 3+ pool, added test case for it

mattverse commented on x/superfluid/keeper/migrate_test.go:
```diff
@@ -831,11 +831,6 @@ func (s *KeeperTestSuite) TestValidateSharesToMigrateUnlockAndExitBalancerPool()
 			overwritePoolId:          true,
 			expectedError:            fmt.Errorf("pool with ID %d does not exist", 2),
 		},
-		"error: attempt to leave a pool that has more than two denoms": {
```
Added back!

p0mvn commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +658,150 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	}
+
+	lockCoin := lock.Coins[0]
+	poolIdLeaving, err := gammtypes.GetPoolIdFromShareDenom(lockCoin.Denom)
```
Can we always get the pool id by using this function? What if this is a CL lock denom?

I'm referring to CL lock denom as made here:
https://github.com/osmosis-labs/osmosis/blob/1bef46333cb2d4b3464f6779e33bea9275e1e41d/x/concentrated-liquidity/position.go#L485

mattverse replied:
AFAIK, we are not supporting CL yet. In that sense, added double check + test case to ensure lock has denom of balancer pool in https://github.com/osmosis-labs/osmosis/pull/5949/commits/be4c03a5c093354abdd82edf83f16dc2d38d5bfd

czarcas7ic replied:
Do we have a test case that attempts to migrate a lock that has a non gamm token prefix?

mattverse replied:
Yes, this is the case
https://github.com/osmosis-labs/osmosis/blob/b953043f91b328699ff62f2f4f4c84c3e978c9c6/x/superfluid/keeper/stake_test.go#L1152-L1157

p0mvn commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,400 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+		expectedError bool
+	}
+	testCases := map[string]tc{
+		"lock that is superfluid delegated": {},
+		"lock that is superfluid undelegating": {
+			unlocking:              true,
+			superfluidUndelegating: true,
+		},
+		"lock that is unlocking": {
+			unlocking:              true,
+			superfluidUndelegating: false,
+		},
+		// error cases
+		"error: min amount to stake greater than actual amount": {
+			useMinAmountToStake: true,
+			expectedError:       true,
+		},
+		"error: sender is not owner of lock ": {
+			senderIsNotOwnerOfLock: true,
+			expectedError:          true,
+		},
```
How are we handling locks migrated to CL? Can we add tests for them as well?

mattverse replied:
ditto for above

p0mvn reviewed (commented):
Left initial set of questions.

Would like to review tests in more detail once feedback is addressed

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+// convertLockToStake handles locks that are superfluid bonded, superfluid unbonding, vanilla locked(unlocking) locks.
+// Deletes all associated state, converts the lock itself to staking delegation by going through exit pool and swap.
+func (k Keeper) convertLockToStake(ctx sdk.Context, sender sdk.AccAddress, valAddr string, lockId uint64,
+	sharesToStake sdk.Coin, minAmtToStake sdk.Int) (totalAmtConverted sdk.Int, shares sdk.Dec, err error) {
```
yes, exactly, changed!

mattverse commented on x/superfluid/types/errors.go:
```diff
@@ -115,3 +117,19 @@ type UnexpectedDenomError struct {
+
+func (e MultipleValFromValsetError) Error() string {
+	return fmt.Sprintf("Multiple validator from valset-preference exists")
+}
```
erased!@

mattverse commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+			s.Require().Error(err)
+
+			// intermediary account should have been deleted in all case
+			if lock.ID != 0 {
+				_, err = s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, lock.ID, keeper.UnstakingSyntheticDenom(lock.Coins[0].Denom, valAddr.String()))
+				s.Require().Error(err)
+			}
```
for the case that is unlocked! (no lock)

mattverse commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+		"error: use different lock ": {
+			usepartialShares: true,
+			expectedError:    true,
+		},
```
Fixed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/0898bdc25c91d561e5f85dac87b4e3dca69431f4

mattverse commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+		"unlocked gamm shares": {
+			unlocking:              true,
+			superfluidUndelegating: false,
+		},
```
Good catch, added and modified test case so that it tests both 
- unlocking case
- unlocked case
in https://github.com/osmosis-labs/osmosis/pull/5949/commits/c89e9c3945e0dc402e836c4805182c1718d81be5

mattverse commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+			s.SetupTest()
+			s.Ctx = s.Ctx.WithBlockTime(defaultJoinTime)
+			// We bundle all migration setup into a single function to avoid repeating the same code for each test case.
+			_, _, lock, _, _, _, _, originalValAddr := s.SetupUnbondConvertAndStakeTest(s.Ctx, true, tc.superfluidUndelegating, false, false)
```
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/c89e9c3945e0dc402e836c4805182c1718d81be5

mattverse commented:
Addressed all comments. 

Major changes are: 

1. Changed so that we delete shares_to_convert (sdk.Coin) field. Effect: liquid gamm share change is now not supported

2. Fall back to superfluid delegation

Upon val address being empty and val set pref delegation not being available, we attempt to fall back to original lock's superfluid delgated address

czarcas7ic commented on x/superfluid/keeper/migrate.go:
```diff
@@ -293,15 +291,15 @@ func (k Keeper) validateMigration(ctx sdk.Context, sender sdk.AccAddress, lockId
 // 5. Ensures that exactly two coins are returned.
 // 6. Any remaining shares that were not migrated are re-locked as a new lock for the remaining time on the lock.
-func (k Keeper) validateSharesToMigrateUnlockAndExitBalancerPool(ctx sdk.Context, sender sdk.AccAddress, poolIdLeaving uint64, lock *lockuptypes.PeriodLock, sharesToMigrate sdk.Coin, tokenOutMins sdk.Coins) (exitCoins sdk.Coins, err error) {
+func (k Keeper) forceUnlockAndExitBalancerPool(ctx sdk.Context, sender sdk.AccAddress, poolIdLeaving uint64, lock *lockuptypes.PeriodLock, sharesToMigrate sdk.Coin, tokenOutMins sdk.Coins, exitCoinsLengthIsTwo bool) (exitCoins sdk.Coins, err error) {
```
Is there any way we can pass in the expected exit coin length and compare that to the length of exit coins instead?

mattverse replied:
I think it might be possible, although it would 1. require a lot of refactoring 2. I think this way is a better way of abstraction. I was thinking this was a better way of abstraction since at the time of function call, we do not know or we do nont have to know how much coin balancer pool is consisted of, but we only have to care about do we want to handle pools consisted of two coins, or do we also want to handle 2+ coin pools. Lmk what your thoughts on this are!

czarcas7ic commented on x/superfluid/keeper/migrate_test.go:
```diff
@@ -797,13 +797,14 @@ func (s *KeeperTestSuite) TestValidateMigration() {
 		overwritePool             bool
 		overwritePoolId           bool
+		exitCoinsLengthIsTwo      bool
 		percentOfSharesToMigrate  sdk.Dec
```
Lets remove percentOfSharesToMigrate as a param now that this is a fixed value.

mattverse replied:
Sorry I don't think I understood you correctly, ForceUnlockAndExitBalancerPool still supports partials right?

czarcas7ic commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +660,183 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+	} else { // liquid gamm shares without locks are not supported
+		return sdk.ZeroInt(), fmt.Errorf("unsupported staking conversion type")
+	}
```
It feels like we could still support this though. We could still take in amount as a param that just gets disregarded if the lockID is non zero.

czarcas7ic commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +660,183 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+		}
+	}
+
+	var val stakingtypes.Validator
```
I don't think these needs to be initialized here

mattverse replied:
Good catch, solved in https://github.com/osmosis-labs/osmosis/pull/5949/commits/0c5b652ee73f06c3fda563502f8a1f8e3b1669b5

czarcas7ic commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1032,454 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+	}
+	testCases := map[string]tc{
+		"lock that is superfluid delegated": {},
+		"lock that is superfluid undelegating": {
+			unlocking:              true,
+			superfluidUndelegating: true,
+		},
+		"bonded lock, not superfluid delegated": {
+			notSuperfluidDelegated: true,
+		},
+		"lock that is unlocking": {
+			unlocking:              true,
+			superfluidUndelegating: true,
+		},
```
Same issue here, `lock that is superfluid undelegating` and `lock that is unlocking` have the same input params, so cant be testing different things.

mattverse replied:
My bad, addressed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/0c5b652ee73f06c3fda563502f8a1f8e3b1669b5

czarcas7ic reviewed (commented):
The core logic appears sound, but still have some comments of items to address. It also feels like the stake_test.go file is really complex and might benefit from logic abstraction.

mattverse commented:
devbot fix

mattverse commented:
@czarcas7ic Attempted to abstract some of the checks we do after tests. If you can point me directions in which part could be better improved for logic abstraction for these test cases, I'd be more than happy to make this cleaner!

p0mvn commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +660,209 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	if migrationType == SuperfluidBonded {
+		_, err = k.undelegateCommon(ctx, sender, lockID)
+		if err != nil {
+			return sdk.ZeroInt(), err
```
ditto for all other instances
```suggestion
			return sdk.Int{}, err
```

mattverse replied:
addressed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/96f486b5e1bbf468406fc4b5d1aa3e17ef6b4003

p0mvn commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +660,209 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	}
+
+	// use routeMigration method to check status of lock (either superfluid staked, superfluid unbonding, vanilla locked, unlocked)
+	_, migrationType, err := k.routeMigration(ctx, senderAddr, int64(lockID))
```
nit: I would rename this to `getMigrationType`. "Route" feels like something that is mutative when this method is not

mattverse replied:
addressed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/96f486b5e1bbf468406fc4b5d1aa3e17ef6b4003

p0mvn commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +660,209 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+
+	// var val stakingtypes.Validator
+	// if given valAddr is empty, we use delegation preference given from valset-pref module or reference from superfluid staking
+	if valAddr == "" {
+		err := k.vspk.DelegateToValidatorSet(ctx, sender.String(), sdk.NewCoin(bondDenom, totalAmtToStake))
+		// if valset-pref delegation errored due to no existing delegation existing, fall back and try using superfluid staked validator
+		if err == valsettypes.ErrNoDelegation {
+			val, err := k.validateValAddrForDelegate(ctx, originalSuperfluidValAddr)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+
+			// delegate now!
+			_, err = k.sk.Delegate(ctx, sender, totalAmtToStake, stakingtypes.Unbonded, val, true)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		} else if err != nil { // for other errors, handle error
+			return sdk.ZeroInt(), err
+		}
+	} else {
+		val, err := k.validateValAddrForDelegate(ctx, valAddr)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+		// delegate now!
+		_, err = k.sk.Delegate(ctx, sender, totalAmtToStake, stakingtypes.Unbonded, val, true)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+	}
```
nit: this could be refactored to reduce code duplication of:

```go
		val, err := k.validateValAddrForDelegate(ctx, valAddr)
		if err != nil {
			return sdk.ZeroInt(), err
		}
		// delegate now!
		_, err = k.sk.Delegate(ctx, sender, totalAmtToStake, stakingtypes.Unbonded, val, true)
		if err != nil {
			return sdk.ZeroInt(), err
		}
```

Let me know if a more specific suggestion could help here

mattverse replied:
addressed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/96f486b5e1bbf468406fc4b5d1aa3e17ef6b4003, also added test cases

p0mvn commented on x/superfluid/types/errors.go:
```diff
@@ -115,3 +117,19 @@ type UnexpectedDenomError struct {
+	return fmt.Sprintf("actual amount converted to stake (%s) is less then minimum amount expected to be staked (%s)", e.ActualTotalAmtToStake, e.ExpectedTotalAmtToStake)
+}
+
+type NoValsetNoSuperfluidDelegationError struct {
+}
+
+func (e NoValsetNoSuperfluidDelegationError) Error() string {
+	return fmt.Sprintf("No val-set for delegation nor existing superfluid delegation")
+}
```
nit: what is the reason for defining the error this way when it has no custom fields (i.e. empty struct)?

mattverse replied:
addressed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/96f486b5e1bbf468406fc4b5d1aa3e17ef6b4003, actually removed the error since it turned out to be unneeded

p0mvn commented on x/superfluid/keeper/stake.go:
```diff
@@ -311,6 +314,7 @@ func (k Keeper) SuperfluidUndelegateToConcentratedPosition(ctx sdk.Context, send
 // partialUndelegateCommon acts similarly to undelegateCommon, but undelegates a partial amount of the lock's delegation rather than the full amount. The amount
 // that is undelegated is placed in a new lock. This function returns the intermediary account associated with the original lock ID as well as the new lock that was created.
 // An error is returned if the amount to undelegate is greater than the locked amount.
+// nolint: unused
 func (k Keeper) partialUndelegateCommon(ctx sdk.Context, sender string, lockID uint64, amountToUndelegate sdk.Coin) (intermediaryAcc types.SuperfluidIntermediaryAccount, newlock *lockuptypes.PeriodLock, err error) {
```
Should we remove this? I'm generally for removing any unused or poorly tested code.

If this is needed in the future, the author can review git history if needed. At the same time, it would grant the opportunity to re-implement the method with fresh invariants in mind.

mattverse replied:
Addressed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/96f486b5e1bbf468406fc4b5d1aa3e17ef6b4003

p0mvn commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,548 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+		unlocked               bool
+		expectedError          error
+	}
+	testCases := map[string]tc{
+		"lock that is superfluid delegated": {},
+		"lock that is superfluid undelegating": {
+			superfluidUndelegating: true,
+		},
+		"bonded lock, not superfluid delegated": {
+			notSuperfluidDelegated: true,
+		},
+		"lock that is unlocking": {
+			unlocking:              true,
+			superfluidUndelegating: true,
+		},
+		"unlocked gamm shares": {
+			notSuperfluidDelegated: true,
+			unlocked:               true,
+		},
```
Do we have a test with CL locks somewhere?

mattverse replied:
Those cases are tested respectively each in TestConvertLockToStake and TestConvertUnlockedToStake

p0mvn replied:
I see this:
https://github.com/osmosis-labs/osmosis/blob/ba1d94388f923880c7a956407855d7c66caf2f4c/x/superfluid/keeper/stake_test.go#L1142-L1148

But I do not see a CL lock explicitly tested.

It would be valuable to set up a separate test case where CL lock is created and we show that converting it fails.

p0mvn reviewed (commented):
@mattverse could you please also add CLI?

It would be great to know which manual localosmosis tests were done prior to merging this

czarcas7ic commented:
All comments I had on latest are more or less nits, except the validate basic comment which is blocking.

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +660,209 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+		}
+	}
+
+	lockCoin := lock.Coins[0]
```
It is done here https://github.com/osmosis-labs/osmosis/blob/a0bfd585ab00277ed087a3c23b29b270a180ab00/x/lockup/types/msgs.go#L46-L47

mattverse commented:
Manual QA results are as following: 
https://hackmd.io/@osmosis/rk2_jQfnh


Basic scenarios are covered in the QA list above, will continue manual QAing for edge cases

p0mvn commented:
> Manual QA results are as following: https://hackmd.io/@osmosis/rk2_jQfnh
> 
> Basic scenarios are covered in the QA list above, will continue manual QAing for edge cases

Could you please also test partial migration + concentrated liquidity lock migration?

p0mvn commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +606,217 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	return totalAmtConverted, nil
+}
+
+// convertGammSharesToOsmoAndStake converts given gamm shares to osmo by swapping in the given pool
+// then stakes it to the designated validator.
+// minAmtToStake works as slippage bound, and would error if total amount being staked is less than min amount to stake.
+// Acutal delegation is done based on input paramters of delegateBaseOnValsetPref.
```
Please also elaborate on parameter values. For example, `originalSuperfluidValAddr` can be an empty string which is unintuitive

mattverse replied:
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed

p0mvn commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +606,217 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	return totalAmtToStake, nil
+}
+
+// delegateBaseOnValsetPref delegates based on given input parameters.
+// Delegation is done in the following logic:
+// - If valAddr provided, single delegate.
+// - If valAddr not provided and valset exists, valsetpref.Delegate
+// - If valAddr not provided and valset delegation is not possible, refer back to original lock's superfluid validator
+// - Else: error
+func (k Keeper) delegateBaseOnValsetPref(ctx sdk.Context, sender sdk.AccAddress, valAddr, originalSuperfluidValAddr string, totalAmtToStake sdk.Int) error {
```
Similar comments about needing to expand on the value of `originalSuperfluidValAddr` and that it can be empty

mattverse replied:
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed

p0mvn reviewed (commented):
I reviewed the core logic and it LGTM so far.

However, I would like to do another cycle prior to approval and after the current comments are addressed, focusing on tests only and actually running them.

ValarDragon commented on proto/osmosis/superfluid/tx.proto:
```diff
@@ -231,4 +236,40 @@ message MsgAddToConcentratedLiquiditySuperfluidPositionResponse {
+    (gogoproto.moretags) = "yaml:\"min_amt_to_stake\"",
+    (gogoproto.nullable) = false
+  ];
+  // shares_to_convert indicates shares wanted to stake.
+  // Note that this field is only used for liquid(unlocked) gamm shares.
+  // For all other cases, this field would be disregarded.
+  cosmos.base.v1beta1.Coin shares_to_convert = 5 [
+    (gogoproto.nullable) = false,
+    (gogoproto.moretags) = "yaml:\"shares_to_convert\"",
+    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coin"
+  ];
```
This does read as maybe we should have two different messages, but perhaps lets save that for a subsequent PR?

mattverse replied:
Yeah tbh I feel the same way, down to make an issue once this get merged

czarcas7ic commented on x/superfluid/client/cli/tx.go:
```diff
@@ -423,3 +424,57 @@ func NewUnlockAndMigrateSharesToFullRangeConcentratedPositionCmd() (*osmocli.TxC
+
+			valAddr := args[1]
+
+			// if user provided args for min amount to stake, use it. If not, use Zero Int.
```
```suggestion
```

czarcas7ic commented on x/superfluid/keeper/stake.go:
```diff
@@ -180,14 +183,17 @@ func (k Keeper) validateLockForSFDelegate(ctx sdk.Context, lock *lockuptypes.Per
 
 // ensure the valAddr is correctly formatted & corresponds to a real validator on chain.
 func (k Keeper) validateValAddrForDelegate(ctx sdk.Context, valAddr string) (stakingtypes.Validator, error) {
+	ctx.Logger().Error("here")
 	valAddress, err := sdk.ValAddressFromBech32(valAddr)
 	if err != nil {
 		return stakingtypes.Validator{}, err
 	}
+	ctx.Logger().Error("here2")
 	validator, found := k.sk.GetValidator(ctx, valAddress)
 	if !found {
 		return stakingtypes.Validator{}, stakingtypes.ErrNoValidatorFound
 	}
+	ctx.Logger().Error("here3")
 	return validator, nil
```
```suggestion
	valAddress, err := sdk.ValAddressFromBech32(valAddr)
	if err != nil {
		return stakingtypes.Validator{}, err
	}
	validator, found := k.sk.GetValidator(ctx, valAddress)
	if !found {
		return stakingtypes.Validator{}, stakingtypes.ErrNoValidatorFound
	}
	return validator, nil
```

czarcas7ic commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +606,217 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+		totalAmtConverted, err = k.convertLockToStake(ctx, senderAddr, valAddr, lockID, minAmtToStake)
+	} else if migrationType == Unlocked {
+		totalAmtConverted, err = k.convertUnlockedToStake(ctx, senderAddr, valAddr, sharesToConvert, minAmtToStake)
+	} else { // liquid gamm shares without locks are not supported
+		return sdk.Int{}, fmt.Errorf("unsupported staking conversion type")
+	}
```
I thought we agreed we were supporting this if the lockID is zero, which is why we added back sharesToConvert

mattverse replied:
Forgot to update comment after adding back sharesToConvert, modified comment to match current status in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed

czarcas7ic commented on x/superfluid/types/msg_test.go:
```diff
@@ -73,3 +75,70 @@ func TestAuthzMsg(t *testing.T) {
 	}
 }
+
+func TestUnbondConvertAndStakeMsg(t *testing.T) {
```
lets test that blank val address works, so it would have caught the bug introduced earlier.

mattverse replied:
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed !

p0mvn commented on x/superfluid/keeper/migrate.go:
```diff
@@ -217,9 +215,9 @@ func (k Keeper) migrateNonSuperfluidLockBalancerToConcentrated(ctx sdk.Context,
 
-// routeMigration determines the status of the provided lock which is used to determine the method for migration.
+// getMigrationType determines the status of the provided lock which is used to determine the method for migration.
 // It also returns the underlying synthetic locks of the provided lock, if any exist.
-func (k Keeper) routeMigration(ctx sdk.Context, sender sdk.AccAddress, providedLockId int64, sharesToMigrate sdk.Coin) (synthLockBeforeMigration lockuptypes.SyntheticLock, migrationType MigrationType, err error) {
+func (k Keeper) getMigrationType(ctx sdk.Context, sender sdk.AccAddress, providedLockId int64) (synthLockBeforeMigration lockuptypes.SyntheticLock, migrationType MigrationType, err error) {
```
@pysel just added it here: https://github.com/osmosis-labs/osmosis/pull/6018

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +606,217 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+	return totalAmtConverted, nil
+}
+
+func (k Keeper) convertUnlockedToStake(ctx sdk.Context, sender sdk.AccAddress, valAddr string, sharesToStake sdk.Coin,
```
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed

mattverse commented on x/superfluid/keeper/stake.go:
```diff
@@ -656,3 +606,217 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+// Delegation is done in the following logic:
+// - If valAddr provided, single delegate.
+// - If valAddr not provided and valset exists, valsetpref.Delegate
+// - If valAddr not provided and valset delegation is not possible, refer back to original lock's superfluid validator
```
That is indeed the case, added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed

mattverse commented:
Addressed all comments, also added requested scenario of 
- partial unlocking 
- trying to stake cl lock 

in hackmd manual QA list

p0mvn commented on .github/workflows/changelog-entry-reminder.yml:
```diff
@@ -14,7 +14,7 @@ jobs:
       - name: Get changed files
         id: changed-files
-        uses: tj-actions/changed-files@v37.6.0
+        uses: tj-actions/changed-files@v37.1.0
```
Looks like a merge conflict here

p0mvn commented on .github/workflows/required_labels.yml:
```diff
@@ -14,7 +14,7 @@ jobs:
       - name: Get changed files
         id: changed-files
-        uses: tj-actions/changed-files@v37.6.0
+        uses: tj-actions/changed-files@v37.1.0
```
and here

p0mvn commented on CHANGELOG.md:
```diff
@@ -44,7 +44,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### API breaks
 
-* [#6014](https://github.com/osmosis-labs/osmosis/pull/6014) refactor: reduce the number of returns in superfluid migration
```
and here

p0mvn commented on app/upgrades/v17/constants.go:
```diff
@@ -355,8 +355,12 @@ func InitializeAssetPairsTestnet(ctx sdk.Context, keepers *keepers.AppKeepers) (
 		spreadFactor := cfmmPool.GetSpreadFactor(ctx)
 		err = validateSpotPriceFallsInBounds(ctx, cfmmPool, keepers, baseAsset, spreadFactor)
-		if err != nil {
+		if errors.Is(err, gammtypes.ErrInvalidMathApprox) {
+			// Result is zero, which means 0.1 osmo was too much for the swap to handle.
+			// This is likely because the pool liquidity is too small, so since this just testnet, we skip it.
 			continue
+		} else if err != nil {
+			return nil, err
```
I suspect that this is also merge conflict

p0mvn commented on app/upgrades/v17/upgrades.go:
```diff
@@ -246,10 +248,14 @@ func flipTwapSpotPriceRecords(ctx sdk.Context, pools []poolmanagertypes.PoolI, k
+			twapRecord.LastErrorTime = time.Time{}
 			oldRecord := twapRecord
+
+			twapRecord.Asset0Denom, twapRecord.Asset1Denom = oldRecord.Asset1Denom, oldRecord.Asset0Denom
```
This as well

p0mvn reviewed (commented):
@mattverse looks like there were merge conflicts. Please take a look

mattverse commented:
Merge conflict situation is pretty bad here 😭

mattverse commented:
yes!!!

czarcas7ic commented:
Just did a quick pass and it appears the merge conflicts were fixed, nice job Matt. Will now do another review, should be my final one, thanks again Matt :)

czarcas7ic commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,677 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+			// gamm check
+			totalPoolLiquidityAfterConvert, err := s.App.GAMMKeeper.GetTotalPoolLiquidity(s.Ctx, poolId)
+			s.Require().NoError(err)
+			// check that pool liquidity have reduced
+			bondDenomPoolAmtAfterConvert := totalPoolLiquidityAfterConvert.AmountOf(bondDenom)
+			s.Require().True(bondDenomPoolAmtAfterConvert.LT(bondDenomPoolAmtBeforeConvert))
```
We should know exactly how much this is reduced by right? Would be nice to check the exact value rather than inequalities if possible.

mattverse replied:
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/fe4b64c8b1e1d22c45d3492539dfeb244ec1fcb0

czarcas7ic reviewed (approved):
Awesome job with all the iterations on this!! Must have been really exhausting :( 

I think we should be looking for two more approvals, but I don't think they will be very far behind as it looks like we have converged on pretty much all comments.

Again, great job :)

mattverse commented:
> Awesome job with all the iterations on this!! Must have been really exhausting :(
> 
> I think we should be looking for two more approvals, but I don't think they will be very far behind as it looks like we have converged on pretty much all comments.
> 
> Again, great job :)

Thanks so much! Thanks for the detailed review iterations as well :) 🙏

p0mvn commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,705 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+
+}
+
+func (s *KeeperTestSuite) delegationCheck(ctx sdk.Context, sender sdk.AccAddress, originalValAddr, newValAddr sdk.ValAddress, totalAmtConverted sdk.Int) {
+	if !originalValAddr.Empty() {
+		// check if original superfluid staked lock's delgation is successfully deleted
+		_, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, originalValAddr)
+		s.Require().False(found)
+	}
+	// check if delegation amount matches
+	delegation, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, newValAddr)
+	s.Require().True(found)
+	s.Require().True(totalAmtConverted.ToDec().Equal(delegation.Shares))
+}
+
+func (s *KeeperTestSuite) lockCheck(ctx sdk.Context, lock lockuptypes.PeriodLock, valAddr string, checkUnstakingSynthLock bool) {
+	// The synthetic lockup should be deleted.
+	_, err := s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, lock.ID, keeper.StakingSyntheticDenom(lock.Coins[0].Denom, valAddr))
+	s.Require().Error(err)
+
+	// intermediary account should have been deleted
+	_, err = s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, lock.ID, keeper.UnstakingSyntheticDenom(lock.Coins[0].Denom, valAddr))
+	s.Require().Error(err)
+
+	// Lock check
+	_, err = s.App.LockupKeeper.GetLockByID(s.Ctx, lock.ID)
+	s.Require().Error(err)
+}
```
Could you please add comments to helpers about what these checks actually validate?

mattverse replied:
Done in https://github.com/osmosis-labs/osmosis/pull/5949/commits/924468690468f7d48e2f52a521bcc297fe4b4792

p0mvn commented on x/superfluid/keeper/stake_test.go:
```diff
@@ -1028,6 +1031,705 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+	s.Require().True(totalAmtConverted.ToDec().Equal(delegation.Shares))
+}
+
+func (s *KeeperTestSuite) lockCheck(ctx sdk.Context, lock lockuptypes.PeriodLock, valAddr string, checkUnstakingSynthLock bool) {
```
The last parameter is unused

mattverse replied:
Done in https://github.com/osmosis-labs/osmosis/pull/5949/commits/924468690468f7d48e2f52a521bcc297fe4b4792

p0mvn reviewed (approved):
Phenomenal work on addressing multiple rounds of feedback and testing everything thoroughly, including localosmosis!

The summaries of localosmosis tests were especially helpful.

I would like to suggest for the future to try adding comment test summaries about the invariants being tested by each test. This would make the test review easier and faster

AlpinYukseloglu reviewed (approved):
LGTM, nice work. No major issues came up on this pass through
